crates/primitives/simulations/src/lib.rs

```diff
@@ -1,16 +1,36 @@
+use blockifier::state::errors::StateError;
+use blockifier::transaction::errors::TransactionExecutionError;
 use blockifier::transaction::objects::TransactionExecutionInfo;
 use starknet_core::types::{SimulationFlag, SimulationFlagForEstimateFee};
 
 // TODO: This is a placeholder
 // https://github.com/starkware-libs/starknet-specs/blob/master/api/starknet_api_openrpc.json#L3919
 // The official rpc expect use to return the trace up to the point of failure.
 // Figuring out how to get that is a problem for later
-#[derive(Debug)]
+#[derive(Debug, Clone)]
 #[cfg_attr(feature = "scale-info", derive(scale_info::TypeInfo))]
 #[cfg_attr(feature = "parity-scale-codec", derive(parity_scale_codec::Encode, parity_scale_codec::Decode))]
-pub struct PlaceHolderErrorTypeForFailedStarknetExecution;
+pub enum SimulationError {
+    ContractNotFound,
+    TransactionExecutionFailed(String),
+    MissingL1GasUsage,
+    FailedToCreateATransactionalStorageExecution,
+    StateDiff,
+}
```

Comment:
I don't think this is correct.

`FailedToCreateATransactionalStorageExecution` is an InternalError. It's our substrate code failing.
While all the others are the users input leading to the starknet protocol being unable to process the transaction.

For the first one we should `log::error` it and return InternalServerError, for the second ones, we should not log it (our code worked as expected, returning an error for an invalid input.

THose are two different levels of errors for two different levels of execution, they should be treated separately.

This is why the `simulation` code return a double error `Result<Result<TransactionExecutionInfo, SimulationError>, SimulationError>`, the outer one is on us, the inner one is on the user. They are managed differently because they are different types of errors